internal/controller/streaming_manager.go

```diff
@@ -192,75 +160,175 @@ func (sm *streamManager) BootUpStreamAndReconnect(ctx context.Context, cancel co
-			responseData, err := am.Onboard(ctx, envMap.TlsSkipVerify, envMap.OnboardingEndpoint)
-			if err != nil {
-				logger.Errorw("Failed to register cluster", "error", err)
-				continue
-			}
-			err = sm.WriteK8sSecret(ctx, responseData, envMap.ClusterCreds)
-			time.Sleep(1 * time.Second)
+			logger.Errorw("Failed to establish stream connection; will retry", "error", err)
+			randomInt, err := rand.Int(rand.Reader, max)
 			if err != nil {
-				am.Logger.Errorw("Failed to write secret to Kubernetes", "error", err)
-			}
-			clientID, clientSecret, err = sm.ReadCredentialsK8sSecrets(ctx, envMap.ClusterCreds)
-			if err != nil {
-				logger.Errorw("Could not read K8s credentials", "error", err)
+				logger.Errorw("Could not generate a random int", "error", err)
 				continue
 			}
+			result := randomInt.Int64()
+			sleep := 1*time.Second + backoff + time.Duration(result)*time.Millisecond // Add randomness
+			time.Sleep(sleep)
+			backoff = backoff * 2 // Exponential increase
+		} else {
+			done <- struct{}{}
+			return
 		}
-		conn, err := SetUpOAuthConnection(ctx, logger, envMap.TokenEndpoint, envMap.TlsSkipVerify, clientID, clientSecret)
+	}
+}
+
+// ExponentialStreamConnect will continue to reboot and restart the main operations within the operator if any disconnects or errors occur.
+func ExponentialStreamConnect(ctx context.Context, logger *zap.SugaredLogger, envMap EnvironmentConfig, bufferedGrpcSyncer *BufferedGrpcWriteSyncer) {
+	for {
+		authConn, client, err := streamAuth(ctx, logger, envMap)
 		if err != nil {
-			logger.Errorw("Failed to set up an OAuth connection", "error", err)
+			logger.Errorw("Failed to establish initial connection; will retry", "error", err)
+			time.Sleep(5 * time.Second) // Retry after a delay
 			continue
 		}
-		sm, err := NewStreams(ctx, logger, conn)
-		if err != nil {
-			logger.Errorw("Failed to create a new stream", "error", err)
-			continue
+
+		instance := &streamClient{
+			conn:   authConn,
+			client: client,
+		}
+
+		sm := &streamManager{
+			instance:           instance,
+			logger:             logger,
+			bufferedGrpcSyncer: bufferedGrpcSyncer,
 		}
 
-		// Update the gRPC client and connection in BufferedGrpcWriteSyncer
-		bufferedGrpcSyncer.UpdateClient(sm.instance.logStream, sm.instance.conn)
-		go bufferedGrpcSyncer.ListenToLogStream()
+		resourceDone := make(chan struct{})
+		logDone := make(chan struct{})
+		sm.bufferedGrpcSyncer.done = logDone
 
-		ctx, cancel := context.WithCancel(ctx)
-		err = sm.BootUpStreamAndReconnect(ctx, cancel)
+		go manageStream(logger, connectAndStreamResources, sm, resourceDone)
+		go manageStream(logger, connectAndStreamLogs, sm, logDone)
+
+		select {
+		case <-resourceDone:
+		case <-logDone:
+		}
+
+		logger.Warn("All streams have been closed. Rebooting the entire connection.")
+	}
+}
+
+func streamAuth(ctx context.Context, logger *zap.SugaredLogger, envMap EnvironmentConfig) (*grpc.ClientConn, pb.KubernetesInfoServiceClient, error) {
```

Comment:
Maybe rename to NewAuthenticatedCon or something? @rlenglet opinions?

Reply:
Yeah, that sounds better.